### PR TITLE
feat(typescript-estree): throw custom error instead of plain object

### LINKED ISSUE #2400: Throw an appropriate Error (an instance of Error class), instead of Object
- [x] I have first restarting my IDE and the issue persists.
- [x] I have updated to the latest version of the packages.
- [x] I have [read the FAQ](https://github.com/typescript-eslint/typescript-eslint/blob/master/docs/getting-started/linting/FAQ.md) and my problem is not listed.

**Repro**

```TS
// the code you're trying to parse
// any invalid TS code
hello {{{{}
```

```TS
// the code you're using to do the parse of the aforementioned code
import { parse, TSESTreeOptions } from '@typescript-eslint/typescript-estree';

export default function (source, options: TSESTreeOptions = {}, defaultOptions: TSESTreeOptions = {
  comment: false,
  jsx: true,
  loc: true,
  range: true
}) {
  return parse(source, {
    ...defaultOptions,
    ...options
  });
}

```

**Expected Result**
Error (that is an instance of Error) being thrown, with stack trace info:
```
{
   message: '';' expected',
   stack: *error stack*,
   // I suggest to extend default Error to add additional params
   details: {
     column: 6,
     index: 43,
     lineNumber: 2,
   }
}
```

**Actual Result**
I'm getting an Object instead of Error
```
{
  index: 43,
  lineNumber: 2,
  column: 6,
  message: "';' expected.",
}
```

**Additional Info**
Throwing anything besides Error is a bad practice because a stack trace is lost.
The stack trace is valuable info.
If it's crucial to add additional info I suggest extending the default Error class with the complementary field and put that info there.

At first glance, it looks like it could be easily done (create extended Error class and modify [createError](https://github.com/typescript-eslint/typescript-eslint/blob/6f397df42cbcf05c10f304c9bbfdae4803aa0ce2/packages/typescript-estree/src/node-utils.ts#L686) method to return `new CustomError({/*object containing details*/}) //error name is a subject to the discussion`)
I would like to submit a PR if no one minds, but I would be glad to hear any critique.

P.S. Of course, I could just wrap all calls to `parse` into `try ... catch` and rethrow, but I would like to see that fixed in-library

**Versions**

| package                                | version |
| -------------------------------------- | ------- |
| `@typescript-eslint/typescript-estree` | `2.34.0` |
| `TypeScript`                           | `3.9.5` |
| `node`                                 | `14.4.0` |

### PR DESCRIPTION
**Pull request by armano2**
throw custom error instead of plain object
additionally expose `fileName` of file that error got reported

I'm unsure if ~`ParseError`~ `TSError` name is correct.

see: https://github.com/Microsoft/TypeScript-wiki/blob/master/Breaking-Changes.md#extending-built-ins-like-error-array-and-map-may-no-longer-work

------------------

note: depends how we look at this change, this can be considered a **breaking change**, even if names of exposed public fields did not change

fixes #2400

### CODE REVIEW

File: packages/typescript-estree/src/error.ts [hunk 1]
```diff
@@ -0,0 +1,17 @@
...
+    public lineNumber: number,
+    public column: number,
+  ) {
+    super(message);
+    Object.setPrototypeOf(this, ParseError.prototype);
+    Object.defineProperty(this, 'name', {
+      value: new.target.name,
+      enumerable: false,
+      configurable: true,
+    });
```
Comment (armano2) [inline packages/typescript-estree/src/error.ts:15]: this is needed to ensure that name of error is set correctly

File: packages/typescript-estree/src/node-utils.ts [hunk 2]
```diff
@@ -668,12 +669,13 @@ export function createError(
   message: string,
 ): TSError {
```
Comment (armano2) [inline packages/typescript-estree/src/node-utils.ts:670]: `TSError` is still present, maybe we should just replace it with new error class?

File: packages/typescript-estree/src/node-utils.ts [hunk 3]
```diff
@@ -649,16 +649,27 @@ export function convertTokens(ast: ts.SourceFile): TSESTree.Token[] {
...
-  lineNumber: number;
-  column: number;
-  message: string;
+export class TSError extends Error {
+  constructor(
+    message: string,
+    public fileName: string,
+    public index: number,
+    public lineNumber: number,
+    public column: number,
```
Comment (bradzacher) [inline packages/typescript-estree/src/node-utils.ts:658]:
```suggestion
    public readonly fileName: string,
    public readonly index: number,
    public readonly lineNumber: number,
    public readonly column: number,
```

File: packages/typescript-estree/src/node-utils.ts [hunk 4]
```diff
@@ -649,16 +649,27 @@ export function convertTokens(ast: ts.SourceFile): TSESTree.Token[] {
...
+  ) {
+    super(message);
+    Object.setPrototypeOf(this, TSError.prototype);
+    Object.defineProperty(this, 'name', {
+      value: new.target.name,
+      enumerable: false,
+      configurable: true,
+    });
```
Comment (bradzacher) [inline packages/typescript-estree/src/node-utils.ts:666]: We target ES2017

https://github.com/typescript-eslint/typescript-eslint/blob/dbe5885c3102d2ce6dddfb37fab681115659bd31/tsconfig.base.json#L18

Which means `super(message)` is not changed and is left as a `super` call.

are you sure we still need this?

**Review by bradzacher — CHANGES_REQUESTED** [review]
do we want to add either a snapshot serialiser so that we can show the line/column in the snapshot output?
This will help us verify things don't regress.

File: packages/typescript-estree/src/node-utils.ts (hunk 4 above)
Comment (armano2) [reply packages/typescript-estree/src/node-utils.ts:666]: yes we do, otherwise name will be Error instead of TSError

Comment (bradzacher) [reply packages/typescript-estree/src/node-utils.ts:666]: can we just make it a readonly property?

What about the `setPrototypeOf` - do we need  that function call at all (that's what I was referring to mostly)

Comment (armano2) [reply packages/typescript-estree/src/node-utils.ts:666]: we kinda do, and its even recommended in ts docs: https://github.com/Microsoft/TypeScript-wiki/blob/master/Breaking-Changes.md#extending-built-ins-like-error-array-and-map-may-no-longer-work

**Comment by armano2** [comment]
> do we want to add either a snapshot serialiser so that we can show the line/column in the snapshot output?
> This will help us verify things don't regress.

simple serializer for this error is added, i'm unsure about format tho

File: packages/typescript-estree/src/node-utils.ts (hunk 4 above)
Comment (bradzacher) [reply packages/typescript-estree/src/node-utils.ts:666]: > we kinda do, and its even recommended in ts docs

But that's in the context of changes in TS2.1's transpilation for ES5 (see the section above it):
https://github.com/Microsoft/TypeScript-wiki/blob/master/Breaking-Changes.md#generated-constructor-code-substitutes-the-return-value-of-super-calls-as-this

As mentioned - we target es2017, so there is no such transpilation:

<details>
<summary>es2017 target</summary>

```js
class TSError extends Error {
    constructor(message, fileName, index, lineNumber, column) {
        super(message); ////////// this is a pure `super` call //////////
        this.fileName = fileName;
        this.index = index;
        this.lineNumber = lineNumber;
        this.column = column;
        Object.setPrototypeOf(this, TSError.prototype);
        Object.defineProperty(this, 'name', {
            value: new.target.name,
            enumerable: false,
            configurable: true,
        });
    }
}
```

</details>

<details>
<summary>es5 target</summary>

```js
var __extends = (this && this.__extends) || (function () {
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (Object.prototype.hasOwnProperty.call(b, p)) d[p] = b[p]; };
        return extendStatics(d, b);
    };
    return function (d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
var TSError = /** @class */ (function (_super) {
    __extends(TSError, _super);
    function TSError(message, fileName, index, lineNumber, column) {
        var _newTarget = this.constructor;
        var _this = _super.call(this, message) || this; ////////// this is not a pure super call //////////
        _this.fileName = fileName;
        _this.index = index;
        _this.lineNumber = lineNumber;
        _this.column = column;
        Object.setPrototypeOf(_this, TSError.prototype);
        Object.defineProperty(_this, 'name', {
            value: _newTarget.name,
            enumerable: false,
            configurable: true,
        });
        return _this;
    }
    return TSError;
}(Error));
```

</details>

File: packages/typescript-estree/tools/tserror-serializer.ts [hunk 5]
```diff
@@ -0,0 +1,14 @@
...
+
+export const serializer: import('pretty-format').Plugin = {
+  test: (val: unknown): val is TSError => val instanceof TSError,
+  serialize(val: TSError, config) {
+    return (
+      `[${val.name}: ${val.message}\n` +
+      `${config.indent}"lineNumber": ${val.lineNumber},\n` +
+      `${config.indent}"column": ${val.column},\n` +
+      `${config.indent}"index": ${val.index},\n` +
+      `]`
```
Comment (bradzacher) [inline packages/typescript-estree/tools/tserror-serializer.ts:11]: I'd just make the format the same as object, but with the new name:

```
TSError {
  "column": ${val.column},
  "index": ${val.index},
  "lineNumber": ${val.lineNumber},
  "message": "${val.message}",
}
```

File: packages/typescript-estree/tools/tserror-serializer.ts [hunk 6]
```diff
@@ -0,0 +1,14 @@
+import { TSError } from '../src/node-utils';
+
+export const serializer: import('pretty-format').Plugin = {
```
Comment (bradzacher) [inline packages/typescript-estree/tools/tserror-serializer.ts:3]: prefer `import type`

Comment (armano2) [reply packages/typescript-estree/tools/tserror-serializer.ts]: ok fixed

File: packages/typescript-estree/src/node-utils.ts (hunk 4 above)
Comment (armano2) [reply packages/typescript-estree/src/node-utils.ts]:
```
Object.setPrototypeOf(this, TSError.prototype);
```
removed

File: packages/typescript-estree/src/node-utils.ts [hunk 7]
```diff
@@ -649,16 +649,26 @@ export function convertTokens(ast: ts.SourceFile): TSESTree.Token[] {
...
-  index: number;
-  lineNumber: number;
-  column: number;
-  message: string;
+export class TSError extends Error {
+  constructor(
+    message: string,
+    public readonly fileName: string,
+    public readonly index: number,
+    public readonly lineNumber: number,
```
Comment (armano2) [inline packages/typescript-estree/src/node-utils.ts:657]: i think we should rename this property, lineNumber is a field present already in Error
```ts
Object.getOwnPropertyNames(new Error())
// [ "fileName", "lineNumber", "columnNumber" ]
```

Comment (armano2) [reply packages/typescript-estree/src/node-utils.ts:657]: what do you think about changing it to

```ts
export interface TSErrorDetails {
  fileName: string;
  index: number;
  lineNumber: number;
  columnNumber: number;
}

export class TSError extends Error {
  constructor(message: string, public readonly details: TSErrorDetails) {
    super(message);
    Object.defineProperty(this, 'name', {
      value: new.target.name,
      enumerable: false,
      configurable: true,
    });
  }
}
```

**Review by bradzacher — APPROVED** [review]

**Pull request merged**